src/pypi_mapping/custom_pypi_mapping.rs

```diff
@@ -0,0 +1,129 @@
+                            .await
+                            .into_diagnostic()
+                            .context("failed to parse pypi name mapping")?;
+
+                        mapping_url_to_name.insert(name.to_string(), mapping_by_name);
+                    }
+                    MappingLocation::Path(path) => {
+                        let contents = std::fs::read_to_string(path)
+                            .into_diagnostic()
+                            .context(format!("mapping on {path:?} could not be loaded"))?;
+                        let data: HashMap<String, String> = serde_json::from_str(&contents)
+                            .unwrap_or_else(|_| {
+                                panic!("Failed to parse JSON mapping located at {path:?}")
+                            });
+
+                        mapping_url_to_name.insert(name.to_string(), data);
+                    }
+                }
+            }
+
+            Ok(mapping_url_to_name)
+        })
+        .await
+}
+
+/// Amend the records with pypi purls if they are not present yet.
+pub async fn amend_pypi_purls(
+    client: &ClientWithMiddleware,
+    mapping_url: &MappingMap,
+    conda_packages: &mut [RepoDataRecord],
+    reporter: Option<Arc<dyn Reporter>>,
+) -> miette::Result<()> {
+    let packages_for_prefix_mapping: Vec<RepoDataRecord> = conda_packages
+        .iter()
+        .filter(|package| !mapping_url.contains_key(&package.channel))
+        .cloned()
+        .collect();
+
+    let prefix_mapping = prefix_pypi_name_mapping::conda_pypi_name_mapping(
+        client,
+        &packages_for_prefix_mapping,
+        reporter,
+    )
+    .await?;
+
+    let custom_mapping = fetch_custom_mapping(client, mapping_url).await?;
+
+    for record in conda_packages.iter_mut() {
+        if !mapping_url.contains_key(&record.channel) {
+            prefix_pypi_name_mapping::amend_pypi_purls_for_record(record, &prefix_mapping)?;
+        } else {
+            amend_pypi_purls_for_record(record, custom_mapping)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Updates the specified repodata record to include an optional PyPI package name if it is missing.
+///
+/// This function guesses the PyPI package name from the conda package name if the record refers to
+/// a conda-forge package.
+fn amend_pypi_purls_for_record(
+    record: &mut RepoDataRecord,
+    custom_mapping: &'static HashMap<String, HashMap<String, String>>,
+) -> miette::Result<()> {
+    // If the package already has a pypi name we can stop here.
+    if record
+        .package_record
+        .purls
+        .iter()
+        .any(|p| p.package_type() == "pypi")
+    {
+        return Ok(());
+    }
+
+    // If this package is a conda-forge package or user specified a custom channel mapping
+    // we can try to guess the pypi name from the conda name
+    if custom_mapping.contains_key(&record.channel) {
+        tracing::warn!("record channel is {}", record.channel);
```

Comment:
Left over debug statement? If not please improve the naming and make it a `debug!` trace

Reply:
removed - thanks